### PR TITLE
init, index: disallow indexes when running reindex-chainstate

### PR DESCRIPTION
**Pull request by mzumsande**
When started together with `-reindex-chainstate`, currently coinstatsindex gets corrupted and the blockfilterindex flatfiles duplicated. See the OP of #24630 for more a more detailed explanation on why this happens.

This is an alternative to #24630 which does not wipe and rebuild the indexes but returns an `InitError` when they are activated, thus requiring the user to deactivate them temporarily until the `-reindex-chainstate` run is finished.

This also disallows `-reindex-chainstate` in combination with `-txindex`, which is not leading to corruption, but currently still rebuilds the index unnecessarily and unexpectedly.

As a long-term goal, it would be desirable to have the indexes tolerate `reindex-chainstate` by ignoring their `BlockConnected` notifications (there is discussion in #24630 about this) or possibly move `reindex-chainstate` option  into a `bitcoin-chainstate` executable, which could also solve the problem. But these would be larger projects - until then, it might be better to disallow the interaction than having corrupted indexes.

The first commit adjusts the `-reindex` doc to mention that this option does rebuild all active indexes.

### CODE REVIEW

**Comment by laanwj** [comment]
Seems ok (with the noted provision that it is a temporary state of affairs). Theoretially `-reindex-chainstate` is supposed to be simply a faster `-reindex` that doesn't rebuild the block index, and work exactly the same otherwise. But if it doesn't work that way, it's better to disable the combination for now.

**Comment by luke-jr** [comment]
Ok with this temporarily, but I think we should [really fix the issue](https://github.com/bitcoin/bitcoin/pull/24630#issuecomment-1079587504) instead long-term.

File: src/init.cpp [hunk 1]
```diff
@@ -1031,6 +1031,19 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
         return InitError(_("No proxy server specified. Use -proxy=<ip> or -proxy=<ip:port>."));
     }
 
+    if (args.IsArgSet("-reindex-chainstate")) {
```
Comment (ryanofsky) [inline src/init.cpp:1034]: In commit "init: disallow reindex-chainstate with optional indexes" (93fa8f651242af5354ba92f7aaec18e878d4f840)

It would be good to replace `args.IsArgSet("-reindex-chainstate")` with `args.GetBoolArg("-reindex-chainstate", false)` because otherwise this will wrongly trigger for `-noreindex-chainstate`.

(In general, it's good to avoid the `IsArgSet` function. It's frequently misused and leads to minor bugs like this.)

File: src/init.cpp [hunk 2]
```diff
@@ -1031,6 +1031,19 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
...
+        // indexes that must be deactivated to prevent index corruption, see #24630
+        if (args.GetBoolArg("-coinstatsindex", DEFAULT_COINSTATSINDEX)) {
+            return InitError(_("reindex-chainstate with active coinstatsindex is not possible. Please deactivate the index temporarily or perform a full -reindex, which will also rebuild the index."));
+        }
+        if (g_enabled_filter_types.count(BlockFilterType::BASIC)) {
+            return InitError(_("reindex-chainstate with active blockfilterindex is not possible. Please deactivate the index temporarily or perform a full -reindex, which will also rebuild the index."));
+        }
+        if (args.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
+            return InitError(_("reindex-chainstate with active txindex is not possible. Please deactivate the index temporarily or perform a full -reindex, which will also rebuild the index."));
```
Comment (ryanofsky) [inline src/init.cpp:1043]: In commit "init: disallow reindex-chainstate with optional indexes" (93fa8f651242af5354ba92f7aaec18e878d4f840)

Curious, but do you know how or whether txindex gets corrupted? #24630 only mentions the other two indexes. I guess even if the txindex doesn't get corrupted, it is probably still good to prevent `-reindex-chainstate` and `-txindex` from being used together if the current implementation is wastefully slow and inefficient, until we properly make the options more compatible.

File: src/init.cpp [hunk 3]
```diff
@@ -1031,6 +1031,19 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
         return InitError(_("No proxy server specified. Use -proxy=<ip> or -proxy=<ip:port>."));
     }
 
+    if (args.IsArgSet("-reindex-chainstate")) {
+        // indexes that must be deactivated to prevent index corruption, see #24630
+        if (args.GetBoolArg("-coinstatsindex", DEFAULT_COINSTATSINDEX)) {
+            return InitError(_("reindex-chainstate with active coinstatsindex is not possible. Please deactivate the index temporarily or perform a full -reindex, which will also rebuild the index."));
```
Comment (ryanofsky) [inline src/init.cpp:1037]: In commit "init: disallow reindex-chainstate with optional indexes" (93fa8f651242af5354ba92f7aaec18e878d4f840)

Minor: Here and below I think you could make recommendations more explicit. I would say "-reindex-chainstate option is not compatible with -coinstatsindex. Please temporarily disable -coinstatsindex while using -reindex-chainstate, or replace -reindex-chainstate with -reindex to fully rebuild all indexes." But current wording is also OK if you prefer it.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 93fa8f651242af5354ba92f7aaec18e878d4f840. This looks great, improves documentation and test coverage, and avoids a corruption bug.

I agree with Luke it would be better in longer term to improve the indexes. Ideally all the indexing options should straightforwardly work well together without duplication, waste, or corruption. But as long as they don't work well together it is good to prevent using them together, and this PR does it in a helpful way, suggesting good workarounds.

It doesn't seem like it should be too hard to follow up this PR with a more permanent fix that just makes indexes ignore blockconnected notifications that aren't relevant or just delays starting them until reindex-chainstate finishes.

File: src/init.cpp (hunk 2 above)
Comment (mzumsande) [reply src/init.cpp:1043]: I don't think that it will get corrupted, because it doesn't have any kind of running variable like the other two indexes. However, it would get "rebuilt" by processing `BlockConnected()` signals for the entire chain (with the result being identical to the previously saved index data), which is probably not be what the user expects from running a `-reindex-chainstate` command. That's why I disabled it as well - I don't think it is strictly necessary and can drop this, if you / others would prefer that.

File: src/init.cpp (hunk 3 above)
Comment (mzumsande) [reply src/init.cpp:1037]: Thanks, I took your suggestion (just didn't use a "-" after "disable" because here the index is meant, not the command option).

File: src/init.cpp (hunk 1 above)
Comment (mzumsande) [reply src/init.cpp:1034]: Done as suggested, good catch! Maybe there could be a warning about this in the doc of `IsArgSet`  (`system.h`).

**Comment by mzumsande** [comment]
Addressed feedback by @ryanofsky

**Review by ryanofsky — APPROVED** [review]
Code review ACK dac44fc06ff9938d070aa5221d106a7f31da9d81. Just fixed IsArgSet call and edited error messages since last review

File: src/init.cpp [hunk 4]
```diff
@@ -1031,6 +1031,19 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
         return InitError(_("No proxy server specified. Use -proxy=<ip> or -proxy=<ip:port>."));
     }
 
+    if (args.GetBoolArg("-reindex-chainstate", false)) {
+        // indexes that must be deactivated to prevent index corruption, see #24630
```
Comment (MarcoFalke) [inline src/init.cpp:1035]: not sure how helpful it is to write `#24630`. Either this should be a full URL or even better a description for devs to explain which indexes must be deactivated and which ones do not.

If all need to be deactivated it would be better to write "All index" instead of "indexes that".

File: src/init.cpp [hunk 5]
```diff
@@ -1031,6 +1031,19 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
...
+        if (args.GetBoolArg("-coinstatsindex", DEFAULT_COINSTATSINDEX)) {
+            return InitError(_("-reindex-chainstate option is not compatible with -coinstatsindex. Please temporarily disable coinstatsindex while using -reindex-chainstate, or replace -reindex-chainstate with -reindex to fully rebuild all indexes."));
+        }
+        if (g_enabled_filter_types.count(BlockFilterType::BASIC)) {
+            return InitError(_("-reindex-chainstate option is not compatible with -blockfilterindex. Please temporarily disable blockfilterindex while using -reindex-chainstate, or replace -reindex-chainstate with -reindex to fully rebuild all indexes."));
+        }
+        if (args.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
+            return InitError(_("-reindex-chainstate option is not compatible with -txindex. Please temporarily disable txindex while using -reindex-chainstate, or replace -reindex-chainstate with -reindex to fully rebuild all indexes."));
```
Comment (MarcoFalke) [inline src/init.cpp:1043]: nit: Maybe this can be wrapped in a lambda that takes the index name?
